storage/postgres/client.go

```diff
@@ -96,89 +95,99 @@ func (c *Client) SendBatch(ctx context.Context, batch *storage.QueryBatch) error
 	return c.SendBatchWithOptions(ctx, batch, pgx.TxOptions{})
 }
 
-func (c *Client) SendBatchWithOptions(ctx context.Context, batch *storage.QueryBatch, opts pgx.TxOptions) error {
-	// NOTE: Sending txs with tx.SendBatch(batch.AsPgxBatch()) is more efficient as it happens
-	// in a single roundtrip to the server.
-	// However, it reports errors poorly: If _any_ query is syntactically
-	// malformed, called with the wrong number of args, or has a type conversion problem,
-	// pgx will report the _first_ query as failing.
-	//
-	// TODO: Remove the first branch if we verify that the performance gain is negligible.
-	if os.Getenv("PGX_FAST_BATCH") == "1" { //nolint:nestif
-		pgxBatch := batch.AsPgxBatch()
-		var batchResults pgx.BatchResults
-		var emptyTxOptions pgx.TxOptions
-		var tx pgx.Tx
-		var err error
-
-		// Begin a transaction.
-		useExplicitTx := opts != emptyTxOptions
-		if useExplicitTx {
-			// set up our own tx with the specified options
-			tx, err = c.pool.BeginTx(ctx, opts)
-			if err != nil {
-				return fmt.Errorf("failed to begin tx: %w", err)
-			}
-			batchResults = c.pool.SendBatch(ctx, &pgxBatch)
-		} else {
-			// use implicit tx provided by SendBatch; see https://github.com/jackc/pgx/issues/879
-			batchResults = c.pool.SendBatch(ctx, &pgxBatch)
+// Submits a new batch. Under the hood, uses `tx.SendBatch(batch.AsPgxBatch())`,
+// which is more efficient as it happens in a single roundtrip to the server.
+// However, it reports errors poorly: If _any_ query is syntactically
+// malformed, called with the wrong number of args, or has a type conversion problem,
+// pgx will report the _first_ query as failing.
+func (c *Client) sendBatchWithOptionsFast(ctx context.Context, batch *storage.QueryBatch, opts pgx.TxOptions) error {
+	pgxBatch := batch.AsPgxBatch()
+	var batchResults pgx.BatchResults
+	var emptyTxOptions pgx.TxOptions
+	var tx pgx.Tx
+	var err error
+
+	// Begin a transaction.
+	useExplicitTx := opts != emptyTxOptions
```

Comment:
should always be a transaction, right? we don't want the batch to be partially applied

Reply:
It is always a tx, but we don't usually need to make it explicit with separate `BEGIN` and `COMMIT` calls. See the comment in line 120 for more context.

Reply:
the comment:

> 		// use implicit tx provided by SendBatch; see https://github.com/jackc/pgx/issues/879

Reply:
ok why do they link to the extended query flow? the section explains,

> In the extended protocol, the frontend first sends a Parse message ...
>
> Note: The query string contained in a Parse message cannot include more than one SQL statement; else a syntax error is reported. This restriction does not exist in the simple-query protocol, but it does exist in the extended protocol, because allowing prepared statements or portals to contain multiple commands would complicate the protocol unduly.

which seems to be rather that there's no grouping of multiple statements into an implicit transaction

by contrast, the simple query section explains:

> When a simple Query message contains more than one SQL statement (separated by semicolons), those statements are executed as a single transaction, unless explicit transaction control commands are included to force a different behavior.

consistent with the summary in the issue comment

which flow does pgx use?

Reply:
pgx uses the extended mode. See [docs](https://github.com/jackc/pgx/blob/7ec6ee7b0ad9ecb4cf885b6630fd36b022e89755/conn.go#L36). I also used wireshark to snoop on the connection, and all PREPARE statements (triggered by both fast and slow tx-sending mode from this PR) contain only a single statement each, which implies extended mode.

In extended mode, pgx always also enables "pipeline mode" ([here](https://github.com/jackc/pgx/blob/7ec6ee7b0ad9ecb4cf885b6630fd36b022e89755/conn.go#L1062)), which "allows applications to send a query without having to read the result of the previously sent query", [per postgres docs](https://www.postgresql.org/docs/current/libpq-pipeline-mode.html). Those same docs also say that "If the pipeline used an implicit transaction(✱), then operations that have already executed are rolled back and operations that were queued to follow the failed operation are skipped entirely," which matches the description by jackc from https://github.com/jackc/pgx/issues/879

✱ Those are statements not inside a BEGIN/COMMIT block. I cannot find an explicit (heh) definition of that exact term, but a [google search](https://www.google.com/search?q=postgres+implicit+transaction) makes the meaning pretty clear, also via official postgres docs.

Reply:
thanks, good find on the pipeline docs saying that

Reply:
philosophically resolved. could you add a code comment about:
- we use pgx cache execution mode
- pgx cache execution mode uses postgresql pipeline
- postgresql pipeline makes an implicit transaction https://www.postgresql.org/docs/current/libpq-pipeline-mode.html#LIBPQ-PIPELINE-SENDING

Reply:
:+1: added